File: archive/examples/python/basic/create_wallet.py
#!/usr/bin/env python3
"""
PAW Blockchain - Create Wallet Example

This example demonstrates how to create a new wallet with a mnemonic phrase.

Usage:
    python create_wallet.py              # Create new wallet
    python create_wallet.py import <mnemonic>  # Import existing wallet

Security Warning:
    Never share your mnemonic phrase or private key with anyone.
    Store them securely offline.
"""

import sys
import hashlib
from typing import Dict, Any, Optional
from mnemonic import Mnemonic
from ecdsa import SigningKey, SECP256k1
from ecdsa.util import sigencode_string_canonize
import bech32

# Wallet configuration
WALLET_PREFIX = 'paw'
HD_PATH = "m/44'/118'/0'/0/0"
MNEMONIC_STRENGTH = 256  # 24 words


class PAWWallet:
    """PAW blockchain wallet"""

    def __init__(self, mnemonic_phrase: str):
        self.mnemonic = mnemonic_phrase
        self.private_key = None
        self.public_key = None
        self.address = None
        self._derive_keys()

    def _derive_keys(self):
        """Derive private and public keys from mnemonic"""
        # Generate seed from mnemonic
        mnemo = Mnemonic("english")
        seed = mnemo.to_seed(self.mnemonic)

        # For simplicity, use first 32 bytes as private key
        # In production, implement full BIP32/BIP44 derivation
        private_key_bytes = seed[:32]

        # Create signing key
        self.private_key = SigningKey.from_string(
            private_key_bytes,
            curve=SECP256k1
        )

        # Get public key
        self.public_key = self.private_key.get_verifying_key()

        # Generate address
        self.address = self._generate_address()

    def _generate_address(self) -> str:
        """Generate bech32 address from public key"""
        # Get compressed public key
        pubkey_bytes = self.public_key.to_string('compressed')

        # Hash public key (SHA256 then RIPEMD160)
        sha256_hash = hashlib.sha256(pubkey_bytes).digest()
        ripemd160_hash = hashlib.new('ripemd160', sha256_hash).digest()

        # Convert to bech32
        converted = bech32.convertbits(ripemd160_hash, 8, 5)
        address = bech32.bech32_encode(WALLET_PREFIX, converted)

        return address

    def get_public_key_hex(self) -> str:
        """Get public key as hex string"""
        return self.public_key.to_string('compressed').hex()


def generate_mnemonic(strength: int = MNEMONIC_STRENGTH) -> str:
    """
    Generate a new mnemonic phrase

    Args:
        strength: Entropy strength (128 = 12 words, 256 = 24 words)

    Returns:
        Mnemonic phrase
    """
    mnemo = Mnemonic("english")
    return mnemo.generate(strength=strength)


def create_wallet(mnemonic: Optional[str] = None) -> Dict[str, Any]:
    """
    Create a new wallet from a mnemonic

    Args:
        mnemonic: Optional mnemonic phrase (generates new if not provided)

    Returns:
        Dict with wallet information
    """
    print('Creating New PAW Wallet...\n')

    try:
        # Generate mnemonic if not provided
        if not mnemonic:
            mnemonic = generate_mnemonic()
            print('✓ Generated new mnemonic phrase')
        else:
            # Validate provided mnemonic
            mnemo = Mnemonic("english")
            if not mnemo.check(mnemonic):
                raise ValueError('Invalid mnemonic phrase')
            print('✓ Using provided mnemonic phrase')

        # Create wallet
        wallet = PAWWallet(mnemonic)

        # Display wallet information
        print('\n' + '=' * 80)
        print('WALLET CREATED SUCCESSFULLY')
        print('=' * 80)
        print('\n⚠️  SECURITY WARNING: Keep this information secure and private!\n')

        print('Mnemonic Phrase (24 words):')
        print('-' * 80)
        print(mnemonic)
        print('-' * 80)

        print('\nWallet Details:')
        print(f'  Address:    {wallet.address}')
        print(f'  Public Key: {wallet.get_public_key_hex()}')
        print(f'  HD Path:    {HD_PATH}')
        print(f'  Prefix:     {WALLET_PREFIX}')

        print('\nNext Steps:')
        print('  1. Save your mnemonic phrase in a secure location')
        print('  2. Never share your mnemonic with anyone')
        print('  3. Fund your wallet with PAW tokens')
        print('  4. Use your address to receive tokens')

        return {
            'success': True,
            'mnemonic': mnemonic,
            'address': wallet.address,
            'public_key': wallet.get_public_key_hex()
        }

    except Exception as e:
        print(f'✗ Error creating wallet: {str(e)}')
        return {
            'success': False,
            'error': str(e)
        }


def import_wallet(mnemonic: str) -> Dict[str, Any]:
    """
    Import existing wallet from mnemonic

    Args:
        mnemonic: Mnemonic phrase

    Returns:
        Dict with wallet information
    """
    print('Importing Existing Wallet...\n')

    try:
        # Validate mnemonic
        mnemo = Mnemonic("english")
        if not mnemo.check(mnemonic):
            raise ValueError('Invalid mnemonic phrase. Please check and try again.')

        # Create wallet
        wallet = PAWWallet(mnemonic)

        print('✓ Wallet imported successfully')
        print(f'\nAddress: {wallet.address}\n')

        return {
            'success': True,
            'address': wallet.address,
            'public_key': wallet.get_public_key_hex()
        }

    except Exception as e:
        print(f'✗ Error importing wallet: {str(e)}')
        return {
            'success': False,
            'error': str(e)
        }


def main():
    """Main entry point"""
    args = sys.argv[1:]

    if len(args) > 0 and args[0] == 'import':
        if len(args) < 2:
            print('Usage: python create_wallet.py import "<mnemonic>"')
            sys.exit(1)
        # Import wallet
        mnemonic = ' '.join(args[1:])
        result = import_wallet(mnemonic)
    else:
        # Create new wallet
        result = create_wallet()

    sys.exit(0 if result['success'] else 1)


if __name__ == '__main__':
    main()

File: archive/examples/javascript/basic/create-wallet.js
#!/usr/bin/env node

/**
 * PAW Blockchain - Create Wallet Example
 *
 * This example demonstrates how to create a new wallet with a mnemonic phrase.
 *
 * Usage:
 *   node create-wallet.js
 *
 * Security Warning:
 *   Never share your mnemonic phrase or private key with anyone.
 *   Store them securely offline.
 */

import { DirectSecp256k1HdWallet } from '@cosmjs/proto-signing';
import { stringToPath } from '@cosmjs/crypto';
import * as bip39 from 'bip39';

// Wallet configuration
const WALLET_PREFIX = 'paw';
const HD_PATH = "m/44'/118'/0'/0/0"; // Standard Cosmos HD path

/**
 * Generate a new mnemonic phrase
 * @param {number} strength - Entropy strength (128 = 12 words, 256 = 24 words)
 * @returns {string} Mnemonic phrase
 */
function generateMnemonic(strength = 256) {
  return bip39.generateMnemonic(strength);
}

/**
 * Create a new wallet from a mnemonic
 * @param {string} mnemonic - Optional mnemonic phrase (generates new if not provided)
 * @returns {Promise<Object>} Wallet information
 */
async function createWallet(mnemonic = null) {
  console.log('Creating New PAW Wallet...\n');

  try {
    // Generate mnemonic if not provided
    if (!mnemonic) {
      mnemonic = generateMnemonic();
      console.log('✓ Generated new mnemonic phrase');
    } else {
      // Validate provided mnemonic
      if (!bip39.validateMnemonic(mnemonic)) {
        throw new Error('Invalid mnemonic phrase');
      }
      console.log('✓ Using provided mnemonic phrase');
    }

    // Create wallet from mnemonic
    const wallet = await DirectSecp256k1HdWallet.fromMnemonic(mnemonic, {
      prefix: WALLET_PREFIX,
      hdPaths: [stringToPath(HD_PATH)]
    });

    // Get accounts from wallet
    const accounts = await wallet.getAccounts();
    const address = accounts[0].address;
    const pubkey = Buffer.from(accounts[0].pubkey).toString('hex');

    // Display wallet information
    console.log('\n' + '='.repeat(80));
    console.log('WALLET CREATED SUCCESSFULLY');
    console.log('='.repeat(80));
    console.log('\n⚠️  SECURITY WARNING: Keep this information secure and private!\n');

    console.log('Mnemonic Phrase (24 words):');
    console.log('-'.repeat(80));
    console.log(mnemonic);
    console.log('-'.repeat(80));

    console.log('\nWallet Details:');
    console.log(`  Address:    ${address}`);
    console.log(`  Public Key: ${pubkey}`);
    console.log(`  HD Path:    ${HD_PATH}`);
    console.log(`  Prefix:     ${WALLET_PREFIX}`);

    console.log('\nNext Steps:');
    console.log('  1. Save your mnemonic phrase in a secure location');
    console.log('  2. Never share your mnemonic with anyone');
    console.log('  3. Fund your wallet with PAW tokens');
    console.log('  4. Use your address to receive tokens');

    return {
      success: true,
      mnemonic,
      address,
      pubkey,
      hdPath: HD_PATH
    };

  } catch (error) {
    console.error('✗ Error creating wallet:', error.message);
    return {
      success: false,
      error: error.message
    };
  }
}

/**
 * Import existing wallet from mnemonic
 * @param {string} mnemonic - Mnemonic phrase
 * @returns {Promise<Object>} Wallet information
 */
async function importWallet(mnemonic) {
  console.log('Importing Existing Wallet...\n');

  try {
    // Validate mnemonic
    if (!bip39.validateMnemonic(mnemonic)) {
      throw new Error('Invalid mnemonic phrase. Please check and try again.');
    }

    const wallet = await DirectSecp256k1HdWallet.fromMnemonic(mnemonic, {
      prefix: WALLET_PREFIX,
      hdPaths: [stringToPath(HD_PATH)]
    });

    const accounts = await wallet.getAccounts();
    const address = accounts[0].address;

    console.log('✓ Wallet imported successfully');
    console.log(`\nAddress: ${address}\n`);

    return {
      success: true,
      address,
      wallet
    };

  } catch (error) {
    console.error('✗ Error importing wallet:', error.message);
    return {
      success: false,
      error: error.message
    };
  }
}

/**
 * Validate a mnemonic phrase
 * @param {string} mnemonic - Mnemonic phrase to validate
 * @returns {boolean} True if valid
 */
function validateMnemonic(mnemonic) {
  try {
    const isValid = bip39.validateMnemonic(mnemonic);
    console.log(`Mnemonic validation: ${isValid ? '✓ Valid' : '✗ Invalid'}`);
    return isValid;
  } catch (error) {
    console.error('✗ Error validating mnemonic:', error.message);
    return false;
  }
}

// Run the example if executed directly
if (import.meta.url === `file://${process.argv[1]}`) {
  const args = process.argv.slice(2);

  if (args[0] === 'import' && args[1]) {
    // Import wallet from provided mnemonic
    importWallet(args.slice(1).join(' '))
      .then(result => {
        if (!result.success) {
          process.exit(1);
        }
      });
  } else if (args[0] === 'validate' && args[1]) {
    // Validate mnemonic
    const isValid = validateMnemonic(args.slice(1).join(' '));
    process.exit(isValid ? 0 : 1);
  } else {
    // Create new wallet
    createWallet()
      .then(result => {
        if (!result.success) {
          process.exit(1);
        }
      });
  }
}

// Export for testing
export { createWallet, importWallet, validateMnemonic, generateMnemonic };
